rp/src/exchange.rs

```diff
@@ -0,0 +1,234 @@
+        let (connection, rtnetlink, _) = rtnetlink::new_connection()?;
+        tokio::spawn(connection);
+
+        // We don't care if this fails, as the device may already have been auto-cleaned up.
+        let _ = rtnetlink.link().del(index).execute().await;
+
+        Ok(())
+    }
+
+    pub async fn wg_set(
+        genetlink: &mut GenetlinkHandle,
+        index: u32,
+        mut attr: Vec<WgDeviceAttrs>,
+    ) -> Result<()> {
+        use futures_util::StreamExt as _;
+        use netlink_packet_core::{NetlinkMessage, NetlinkPayload};
+        use netlink_packet_generic::GenlMessage;
+        use netlink_packet_wireguard::{Wireguard, WireguardCmd};
+
+        attr.insert(0, WgDeviceAttrs::IfIndex(index));
+
+        let wgc = Wireguard {
+            cmd: WireguardCmd::SetDevice,
+            nlas: attr,
+        };
+
+        let genl = GenlMessage::from_payload(wgc);
+        let mut nlmsg = NetlinkMessage::from(genl);
+        nlmsg.header.flags = NLM_F_REQUEST | NLM_F_ACK;
+
+        let (res, _) = genetlink.request(nlmsg).await?.into_future().await;
+        if let Some(res) = res {
+            let res = res?;
+            if let NetlinkPayload::Error(err) = res.payload {
+                return Err(err.to_io().into());
+            }
+        }
+
+        Ok(())
+    }
+}
+
+#[cfg(any(target_os = "linux", target_os = "freebsd"))]
+pub async fn exchange(options: ExchangeOptions) -> Result<()> {
+    use std::fs::{self, read_to_string};
+
+    use anyhow::anyhow;
+    use base64::Engine;
+    use netlink_packet_wireguard::{constants::WG_KEY_LEN, nlas::WgDeviceAttrs};
+    use rosenpass::{
+        app_server::{AppServer, WireguardOut},
+        config::Verbosity,
+        protocol::{SPk, SSk, SymKey},
+    };
+    use rosenpass_util::file::{LoadValue as _, LoadValueB64};
+    use zeroize::Zeroize;
+
+    let (connection, rtnetlink, _) = rtnetlink::new_connection()?;
+    tokio::spawn(connection);
+
+    let link_name = options.dev.unwrap_or("rosenpass0".to_string());
+    let link_index = netlink::link_create_and_up(&rtnetlink, link_name.clone()).await?;
+
+    ctrlc_async::set_async_handler(async move {
+        netlink::link_cleanup_standalone(link_index)
+            .await
+            .expect("Failed to clean up");
+    })?;
+
+    // Deploy the classic wireguard private key
+    let (connection, mut genetlink, _) = genetlink::new_connection()?;
+    tokio::spawn(connection);
+
+    let wgsk_path = options.private_keys_dir.join("wgsk");
+    let mut wgsk: [u8; WG_KEY_LEN] = base64::engine::general_purpose::STANDARD
+        .decode(&read_to_string(wgsk_path)?)?
+        .try_into()
+        .map_err(|_| anyhow!("WireGuard secret key is not {} bytes long.", WG_KEY_LEN))?;
+
+    let mut attr: Vec<WgDeviceAttrs> = Vec::with_capacity(2);
+    attr.push(WgDeviceAttrs::PrivateKey(wgsk));
+
+    if let Some(listen) = options.listen {
+        attr.push(WgDeviceAttrs::ListenPort(listen.port() + 1));
```

Comment:
I think we have to handle the edge case of a uint16 overflow here. What happens if `listen.port()` equals 65535? I think it might be better to catch this in CLI parsing though, but maybe still add an assertion here. :-)

Reply:
Rust will panic complaining about integer overflow. I can add a nicer error message.

Reply:
Resolved in 679b67a228117f79e4184dac7b6f403c2bed47ca.